### ISSUE TITLE
can Jayson do "bidirectional" calling over websockets?

### ISSUE THREAD

**Issue by darrencruse** [tedeh/jayson#216]
hi really just a question:

Using jayson can a browser (client) make an rpc call to the (node.js) server and then the server make an rpc call back to the browser client?

if so is there an example of that anywhere?

otherwise fwiw I do see some other json-rpc libraries showing examples of this ability when using websockets.

I didn't see a way to add handlers for rpc calls on the *client* object in the jayson api, but Is it conceivable an initial websocket connection could be established with browser as client and then would/could it be the node.js server could create a jayson *client* instance and the browser create a jayson *server* instance passing in that websocket connection?

i.e. to enable calls going in both directions?

**Comment by damartripamungkas** [comment]
maybe you mean you want to do a webhook but use a websocket?

**Comment by damartripamungkas** [comment]
> hi really just a question:
> 
> Using jayson can a browser (client) make an rpc call to the (node.js) server and then the server make an rpc call back to the browser client?
> 
> if so is there an example of that anywhere?
> 
> otherwise fwiw I do see some other json-rpc libraries showing examples of this ability when using websockets.
> 
> I didn't see a way to add handlers for rpc calls on the _client_ object in the jayson api, but Is it conceivable an initial websocket connection could be established with browser as client and then would/could it be the node.js server could create a jayson _client_ instance and the browser create a jayson _server_ instance passing in that websocket connection?
> 
> i.e. to enable calls going in both directions?

maybe you mean you want to do a webhook but use a websocket?

**Comment by tedeh** [comment]
Yes I guess it is possible. You would have both JSON-RPC requests and JSON-RPC responses coming in the message handler on both the server and client side. You need to keep track of outstanding requests so that you can pair a sent request with a corresponding incoming response, and handle timeouts for this.

Lots of opportunities for deadlocks.

Unfortunately I don't have time to write an example.

**Comment by tedeh** [comment]
Actually I did make a quite simple example of a duplex websocket here: https://github.com/tedeh/jayson/tree/master/examples/websocket_duplex
